### PR TITLE
feat: add Windows support with the `nil` backend

### PR DESCRIPTION
**Pull request by haraldh**
Also support `test --workspace` on MacOS and Windows.

Signed-off-by: Harald Hoyer <harald@profian.com>

### CODE REVIEW

File: crates/exec-wasmtime/src/lib.rs [hunk 1]
```diff
@@ -43,17 +42,53 @@ pub enum Package {
...
+    let attested = requested.next()?;
+    let compiled = attested.next()?;
+    let connected = compiled.next()?;
+    let completed = connected.next()?;
+    drop(completed);
+    Ok(())
+}
+
+#[cfg(unix)]
 /// Execute
```
Comment (rvolosatovs) [inline crates/exec-wasmtime/src/lib.rs:85]: Let's document that this assumes a Unix socket on fd 3 and reads and parses args from there.

File: src/main.rs [hunk 2]
```diff
@@ -167,23 +189,36 @@ fn main() -> Result<()> {
...
+                    vec![wasm].into_iter().chain(conf),
+                ))
+            };
+
+            #[cfg(windows)]
+            let get_pkg = || {
+                let (wasm, conf) = open_package(module, wasmcfgfile)?;
+                Ok((
+                    Package::Local { wasm, conf },
+                    empty(), // vec![wasm].into_iter().chain(conf)
```
Comment (rvolosatovs) [inline src/main.rs:210]: What's the commented-out code for?

File: src/main.rs [hunk 3]
```diff
@@ -238,7 +275,18 @@ fn main() -> Result<()> {
...
                         ))
-                    })?
+                    };
+
+                    #[cfg(windows)]
+                    let get_pkg = || {
+                        let (wasm, conf) = open_package(wasm, conf)?;
+                        Ok((
+                            Package::Local { wasm, conf },
+                            empty(), // vec![wasm].into_iter().chain(conf)
```
Comment (rvolosatovs) [inline src/main.rs:285]: see above

File: src/main.rs [hunk 4]
```diff
@@ -238,7 +275,18 @@ fn main() -> Result<()> {
                             },
                             vec![wasm].into_iter().chain(conf),
```
Comment (rvolosatovs) [inline src/main.rs:276]: Couldn't we do the config switch on this line instead to reduce duplication?

**Review by rvolosatovs — COMMENTED** [review]
Why don't we use something like https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipes instead, or better yet an `AF_UNIX` socket (I looked through Rust stdlib docs, but it seems that Windows support here is not straight-forward).
See also https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/

File: src/main.rs (hunk 4 above)
Comment (haraldh) [reply src/main.rs]: done

File: crates/exec-wasmtime/src/lib.rs (hunk 1 above)
Comment (haraldh) [reply crates/exec-wasmtime/src/lib.rs:85]: done

File: crates/exec-wasmtime/src/lib.rs [hunk 5]
```diff
@@ -43,17 +42,53 @@ pub enum Package {
...
+#[derive(Debug)]
+pub enum Package {
+    /// Remote URL to fetch package from
+    Remote(Url),
+
+    /// Local package
+    Local {
+        /// Open WASM module file Handle
+        wasm: std::fs::File,
+        /// Optional open config file descriptor
```
[the remark this answers is not in the document]
Comment (haraldh) [reply crates/exec-wasmtime/src/lib.rs]: done

File: src/main.rs (hunk 2 above)
Comment (haraldh) [reply src/main.rs]: removed it .. see below

**Comment by haraldh** [comment]
Although I created the uds_windows crate, I think windows is lacking socketpair and abstract unix sockets. While we could use temp files and pass the filename, I like the direct File approach more

**Review by bstrie — COMMENTED** [review]
Looks pretty good, although in general I'm worried that we're growing a lot of conditional-compilation profiles that are going to be hard to keep track of.

**Comment by haraldh** [comment]
Although I created the [uds_windows crate](https://crates.io/crates/uds_windows), I like the direct File approach more.

Because windows is lacking socketpair and abstract unix sockets, we would have to use named pipes, which would mean different structs anyway.

And still we would have to serialize things though serde_json... which would mean to transform `Handle` -> `(void *)` to `usize` and then back again, which is more ugly than what I propose here I guess.

File: crates/exec-wasmtime/src/lib.rs [hunk 6]
```diff
@@ -13,12 +13,10 @@ mod loader;
...
 use url::Url;
 
-use std::io::Read;
-use std::mem::forget;
+#[cfg(unix)]
 use std::os::unix::io::{FromRawFd, RawFd};
-use std::os::unix::net::UnixStream;
 
-use anyhow::Context;
+#[cfg(unix)]
```
[the remark this answers is not in the document]
Comment (haraldh) [reply crates/exec-wasmtime/src/lib.rs:19]:
```
53   | /         /// Open WASM module file
54   | |         wasm: std::fs::File,
     | |___________________________^ the trait `Deserialize<'_>` is not implemented for `std::fs::File`
```

[the remark this answers is not in the document]
Comment (haraldh) [reply crates/exec-wasmtime/src/lib.rs:19]: > And while we're at it, did you want to make serde a conditional dependency in Cargo.toml?

done ... [8c58a1d](https://github.com/enarx/enarx/pull/1890/commits/8c58a1d967ff192c352195de12fa5253696120c7)

[the remark this answers is not in the document]
Comment (haraldh) [reply crates/exec-wasmtime/src/lib.rs:19]: oops, that didn't work out :) Of course it is still used.... removed again

[the remark this answers is not in the document]
Comment (haraldh) [reply crates/exec-wasmtime/src/lib.rs:19]: see https://github.com/enarx/enarx/runs/6503206014?check_suite_focus=true

**Comment by haraldh** [comment]
> Although I created the [uds_windows crate](https://crates.io/crates/uds_windows), I like the direct File approach more.
> 
> Because windows is lacking socketpair ....

@rvolosatovs @bstrie 

I stand corrected... someone with the name "Harald Hoyer" [added some undocumented method `pair()`](https://github.com/haraldh/rust_uds_windows/commit/1d4073d3f237400a7509721c6b81162a216e06a7) on Dec 12 2018 ... **cough**

https://docs.rs/uds_windows/1.0.2/uds_windows/struct.UnixStream.html#method.pair

Let me revise my strategy and maybe revert this PR partially... :) `RawSocket` is `u64`, so we could merge `RawFd` and `RawSocket` in a u64.

**Pull request merged**